### PR TITLE
[TKW] Read type fixes

### PR DESCRIPTION
**Pull request by Hardcode84**
* Change `Read` return type from from `Memory` to `Register` and make it respect mapping shape. 
* Add `memory_type` property to get the original mem type
* Update related passes
* Some smaller fixes:
  * `_contains` func to properly handle case when input is `None`
  * `safe_subs` to properly handle case when input is not a sympy expr (e.g. when `address_space` has explicit value set instead of sympy placeholder)
  * `custom.axis.name` -> `str(custom.axis)` to handle cases when axis is expr and not just a symbol

### CODE REVIEW

File: shark_turbine/kernel/ops/wave_ops.py [hunk 1]
```diff
@@ -765,9 +765,11 @@ def post_expansion(self, constraints: list["Constraint"]) -> None:
         self.rhs.index = self.rhs_index
         self.acc.index = self.acc_index
 
-        if self.lhs_type.address_space == SHARED_ADDRESS_SPACE:
+        # TODO: this is really wrong place for it, it relies on specific kernel structure,
+        # generated, if mma input is not come from load, things will breaks.
```
Comment (harsh-nod) [inline shark_turbine/kernel/ops/wave_ops.py:769]: Yes good point. With the separation into type and memory_type, not all ops will have a memory_type and so this won't work. Can you move this into the index setter for Read and Write Ops which I think will be a better place for this? So the idea is that we set the lhs, rhs and acc indices here. But within the Read and Write index setters, we can check if the memory type has a shared address space and if so we remove the global indexing.

File: shark_turbine/kernel/wave/utils.py [hunk 2]
```diff
@@ -305,3 +305,14 @@ def compile_and_invoke(
             **benchmark_flags,
         )
         _print_bench_result(benchmark_results, bench_file)
+
+
+def safe_subs(input: Any, subs: List[Tuple[IndexExpr, IndexExpr]]) -> Any:
+    """
+    Substitute input using provided `subs` list if input is sympy object.
+    Otherwise return input unchanged.
+    """
```
Comment (harsh-nod) [inline shark_turbine/kernel/wave/utils.py:314]: Nice! Should we instantiate the indexing context within this function so you don't need to pass subs or you can make subs optional? That will avoid having to instantiate an indexing context multiple times in the code.

**Review by harsh-nod — CHANGES_REQUESTED** [review]
Nice! Thanks for the cleanup. Just a comment about where to put the code to remove global indexing.

**Review by raikonenfnu — COMMENTED** [review]
Aside from Harsh's Qs, looks good to me!

File: shark_turbine/kernel/wave/utils.py (hunk 2 above)
Comment (Hardcode84) [reply shark_turbine/kernel/wave/utils.py:314]: Done

File: shark_turbine/kernel/ops/wave_ops.py (hunk 1 above)
Comment (Hardcode84) [reply shark_turbine/kernel/ops/wave_ops.py:769]: I tried to make index setter smart, but it didn't wok., `remove_global_indexing` need access to constraints, which are not available in setter.

**Comment by Hardcode84** [comment]
+ MI300 CI seems to be broken, working fine locally on my snapshot (IREE version update?)

File: shark_turbine/kernel/ops/wave_ops.py (hunk 1 above)
Comment (harsh-nod) [reply shark_turbine/kernel/ops/wave_ops.py:769]: Thanks for trying, I can take a look at this.

**Review by harsh-nod — APPROVED** [review]
lgtm! thanks!

**Comment by harsh-nod** [comment]
Usually re-triggering the job works but we should take a closer look at this. @saienduri

**Pull request merged**